rebase onto v1.2.0-beta.1

liggitt opened the pull request:
- [x] Rebase https://github.com/openshift/kubernetes/tree/stable-20160309 onto https://github.com/kubernetes/kubernetes/tree/v1.2.0-beta.1 (see https://github.com/openshift/kubernetes/tree/release-1.2-stable-20160312)
- [x] Bump k8s.io/kubernetes Godep to https://github.com/openshift/kubernetes/commit/148dd34ab0e7daeb82582d6ea8e840c15a24e745
- [x] Origin build
- [x] unit tests
- [x] test-cmd
- [x] integration tests
- [x] e2e
- [x] check open ports
- [x] review policy changes for autoscaling and batch API groups
- [x] Rework dropped commits:
  - [x] 378f6f1f1e23c61d5ed4321b0ff7241ccc111867 `UPSTREAM: <carry>: Keep default generator for run 'run-pod/v1'` - replace with oc-specific overrides
  - [x] 055d8eccf9942063113b174b98535f82618327db `UPSTREAM: 12221: Allow custom namespace creation in e2e framework` - replaced by 20808 upstream
  - [x] e4d9fb6906dbef5a069e5f45ee64a49d20418c2e `UPSTREAM: 20706: register internal types with scheme for reference unit test` - replaced with conditional re-register
- [x] `should KubeletCgroups and SystemCgroups also be set to ""?`
- [x] `right default for defaultReplenishmentSyncPeriod?`
- [x] resolve policy reconcile for discovery with namespace controller (discovery rules live in the StatusChecker role which isn't autoreconciled on startup)

Follow-ups:
- [ ] p1: recheck extended tests (current status against master: https://github.com/openshift/origin/issues/6882#issuecomment-194487536)
- [ ] p1: Allow storage versions for API groups to be configurable (see upstream)
- [ ] p1: Set up cross-encoding storage? (see upstream... not sure why this is needed)
- [ ] p1: sweep upstream config/initialization to make sure we match
  - [ ] kube-apiserver
  - [ ] kubelet
  - [ ] kube-proxy
  - [ ] kube-controllermanager
- [ ] p2: fix slow etcd unit tests and lower unit test timeout again
    ```
github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/storage/etcd	45.833s	coverage: 70.2% of statements
github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/registry/generic/etcd	28.310s	coverage: 57.5% of statements
    ```

- [ ] p2: update node_auth to use nodes/proxy subresource in SAR, so we can gain verb info? 
- [ ] p2: update callers to handle partial success from reconcile-cluster-role-bindings
- [ ] p2: determine a preferred GVK to pass to `InfoForObject` in bulk create?
- [ ] p2: `ServerAddressByClientCIDRs: s.getServerAddressByClientCIDRs(req.Request) in discovery doc`
- [ ] p2: Enable client generation for our types
- [ ] p2: Start using genericapiserver framework
- [ ] p2: Sweep for use of kapi.RestMapper
- [ ] p2: Start using the generic namespace controller
- [ ] p3: remove authorizationapi groups
- [ ] p3: improve factoring of volume config upstream to allow passing volume config through (c.f. https://github.com/liggitt/origin/commit/5e9855e1768e84f48a03b7a4e5a7e76a8c471c6f#commitcomment-16705544)
- [ ] p3: upstream issue for job/hpa/daemonset/replicaset controllers to use discovery, support batch/autoscaling API calls if extensions is disabled

liggitt commented:
[test]

liggitt commented:
Best flake ever... this worked until DST hit:
```
--- FAIL: TestTranslateTimestamp (0.00s)
	resource_printer_test.go:1265: On 2 days ago, expected '2d', but got '1d'
	resource_printer_test.go:1265: On months ago, expected '90d', but got '89d'
FAIL
coverage: 47.8% of statements
FAIL	github.com/openshift/origin/Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl	2.725s
```

smarterclayton commented:
Will have to dig into the int test tomorrow

liggitt commented:
e2e makes it to this line:
```
# Wait for build which should have triggered automatically
echo "[INFO] Starting build from ${STI_CONFIG_FILE} and streaming its logs..."
#oc start-build -n test ruby-sample-build --follow
os::build:wait_for_start "test"
# Ensure that the build pod doesn't allow exec
[ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
os::build:wait_for_end "test"
```
and the build fails with this error:
```
F0314 14:04:09.180362       1 builder.go:204] Error: build error: start container "openshift_s2i-build_ruby-sample-build-1_test_post-commit": API error (500): Cannot start container 1289cab6d3a1794cd52ce64239cb9608503c7903a95f0a342edd70c987c21f6c: [8] System error: open
 /sys/fs/cgroup/cpu,cpuacct/system.slice/docker-1289cab6d3a1794cd52ce64239cb9608503c7903a95f0a342edd70c987c21f6c.scope/cpu.cfs_quota_us: no such file or directory
```

smarterclayton commented:
That's the flake - @mrunalp ^ if we need more eyes here to chase these down
I think these are things we need to have fixed.

On Mon, Mar 14, 2016 at 10:13 AM, Jordan Liggitt <notifications@github.com>
wrote:

> e2e makes it to this line:
>
> # Wait for build which should have triggered automatically
> echo "[INFO] Starting build from ${STI_CONFIG_FILE} and streaming its logs..."
> #oc start-build -n test ruby-sample-build --follow
> os::build:wait_for_start "test"
> # Ensure that the build pod doesn't allow exec
> [ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
> os::build:wait_for_end "test"
>
> and the build fails with this error:
>
> F0314 14:04:09.180362       1 builder.go:204] Error: build error: start container "openshift_s2i-build_ruby-sample-build-1_test_post-commit": API error (500): Cannot start container 1289cab6d3a1794cd52ce64239cb9608503c7903a95f0a342edd70c987c21f6c: [8] System error: open
>  /sys/fs/cgroup/cpu,cpuacct/system.slice/docker-1289cab6d3a1794cd52ce64239cb9608503c7903a95f0a342edd70c987c21f6c.scope/cpu.cfs_quota_us: no such file or directory
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7975#issuecomment-196324686>.
>

mrunalp commented:
Clayton this one should be fixed. I will check on the rpm update.

Sent from my iPhone

> On Mar 14, 2016, at 9:08 AM, Clayton Coleman <notifications@github.com> wrote:
> 
> That's the flake - @mrunalp ^ if we need more eyes here to chase these down
> I think these are things we need to have fixed.
> 
> On Mon, Mar 14, 2016 at 10:13 AM, Jordan Liggitt <notifications@github.com>
> wrote:
> 
> > e2e makes it to this line:
> >
> > # Wait for build which should have triggered automatically
> > echo "[INFO] Starting build from ${STI_CONFIG_FILE} and streaming its logs..."
> > #oc start-build -n test ruby-sample-build --follow
> > os::build:wait_for_start "test"
> > # Ensure that the build pod doesn't allow exec
> > [ "$(oc rsh ${BUILD_ID}-build 2>&1 | grep 'forbidden')" ]
> > os::build:wait_for_end "test"
> >
> > and the build fails with this error:
> >
> > F0314 14:04:09.180362 1 builder.go:204] Error: build error: start container "openshift_s2i-build_ruby-sample-build-1_test_post-commit": API error (500): Cannot start container 1289cab6d3a1794cd52ce64239cb9608503c7903a95f0a342edd70c987c21f6c: [8] System error: open
> > /sys/fs/cgroup/cpu,cpuacct/system.slice/docker-1289cab6d3a1794cd52ce64239cb9608503c7903a95f0a342edd70c987c21f6c.scope/cpu.cfs_quota_us: no such file or directory
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/openshift/origin/pull/7975#issuecomment-196324686>.
> >
> —
> Reply to this email directly or view it on GitHub.
>

mrunalp commented:
@smarterclayton There are two classes of issues:
1. Like this one and in #7616 where we don't see some cgroups files. https://github.com/projectatomic/docker/pull/71 is a fix for it and should be available in a rpm soon. 
2. cgroup.procs missing like https://github.com/openshift/origin/issues/7927. I have pinged systemd devs on that and will update that issue once I hear back. (In this case, the cgroup.proc file goes missing from under the libcontainer code that actually creates the file and tries to write to it and I suspect systemd to be removing it.)

smarterclayton commented:
Thanks for the links.

On Mon, Mar 14, 2016 at 1:09 PM, Mrunal Patel <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> There are two classes
> of issues:
> 1. Like this one and in #7616
> <https://github.com/openshift/origin/issues/7616> where we don't see some
> cgroups files. projectatomic/docker#71
> <https://github.com/projectatomic/docker/pull/71> is a fix for it and
> should be available in a rpm soon.
> 2. cgroup.procs missing like #7927
> <https://github.com/openshift/origin/issues/7927>. I have pinged systemd
> devs on that and will update that issue once I hear back. (In this case,
> the cgroup.proc file goes missing from under the libcontainer code that
> actually creates the file and tries to write to it and I suspect systemd to
> be removing it.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7975#issuecomment-196415367>.
>

liggitt commented:
@smarterclayton, looks like TestImageStreamImportAuthenticated is broken because of effectivePath calculation in `Godeps/_workspace/src/k8s.io/kubernetes/pkg/credentialprovider/keyring.go` ignores the original path if it starts with `/v2/` (see discussion in https://github.com/kubernetes/kubernetes/pull/21286/files#r53490530). In addition to the effective path computation, we also need to include the original specified path.

liggitt commented:
@smarterclayton review https://github.com/openshift/origin/commit/250eb94573e10db8f33e8adc09dc00a49e96abdd

liggitt commented:
woohoo!

> continuous-integration/openshift-jenkins/test SUCCESS (https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/2126/)

liggitt commented:
[testonlyextended][extended:all]

liggitt commented:
[test][extended:all]

liggitt commented:
clean unit, cmd, integration, and e2e runs. kicking off extended tests, a few questions marked as TODOs, but this is ready for review

liggitt commented:
feel free to look at whatever, but I'd especially like eyes on these areas:

@derekwaynecarr - defaultReplenishmentSyncPeriod question, general namespace controller wiring, cgroup question at https://github.com/liggitt/origin/commit/5e9855e1768e84f48a03b7a4e5a7e76a8c471c6f#diff-cad467eadf0e2416f7628c1892e4e785 (or maybe @marun might know for DIND mode)
@deads2k - storage version config, restmapper changes, API group policy changes
@soltysh - extended test namespace creation
@smarterclayton - run generator override, imageimportwithauth test fix
@markturansky - PV controller config hookup

marun commented:
@liggitt So long as the networking extended tests can pass on rhel (the ci default), this lgtm.

marun commented:
re-[test]

same aws provisioning hiccups as yesterday morning.

marun commented:
@liggitt ci is failing due to golang syntax error.

    Stderr from the command:

    ./pkg/cmd/util/clientcmd/factory_test.go:25:118: expected ';', found ']'

liggitt commented:
comments addressed, configurable storage versions moved to followup.
still need to figure out discovery reconcile at startup for namespace controller, and look at the InfoForObject we pass during bulk create.
other than that, we'll look at the extended test results overnight and hopefully squash and merge tomorrow

soltysh commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/apis/extensions/validation/validation.go:
```diff
@@ -419,6 +419,10 @@ func ValidateGeneratedSelector(obj *extensions.Job) field.ErrorList {
 	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), "controller-uid", string(obj.UID))...)
 	allErrs = append(allErrs, apivalidation.ValidateHasLabel(obj.Spec.Template.ObjectMeta, field.NewPath("spec").Child("template").Child("metadata"), "job-name", string(obj.Name))...)
 	expectedLabels := make(map[string]string)
+	// Expect all the labels that are actually on the template
```
Since there's no clear agreement on this upstream issue, I'd rather drop this. My guts tell me we might end-up with a nasty carry.

liggitt replied:
Sure, waiting to hear how I'm supposed to actually use the API to accomplish what we used to do

soltysh commented:
@liggitt since github does not like me today (for some unknown reason) and hides my comment I'm gonna put it here, so it's not missed. Since there's no clear agreement on [UPSTREAM: 22921: Fix job selector validation and tests](https://github.com/openshift/origin/pull/7975/commits/c7bb75907fa0687e8667fdc782724cb70041f62b), I'd rather drop this. My guts tell me we might end-up with a nasty carry.

soltysh commented on pkg/scheduler/admission/podnodeconstraints/admission.go:
```diff
@@ -59,6 +60,7 @@ var resourcesToCheck = map[unversioned.GroupResource]unversioned.GroupKind{
 	kapi.Resource("pods"):                   kapi.Kind("Pod"),
 	kapi.Resource("podtemplates"):           kapi.Kind("PodTemplate"),
 	kapi.Resource("replicationcontrollers"): kapi.Kind("ReplicationController"),
+	batch.Resource("jobs"):                  batch.Kind("Job"),
```
`getPodSpec` needs update as well with batch.Job.

liggitt replied:
Nope, there is no internal batch.Jobs object

soltysh replied:
Ah, ok, it's working against internal object, my bad, sorry.

soltysh commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/run.go:
```diff
@@ -148,6 +156,14 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 	}
 
 	generatorName := cmdutil.GetFlagString(cmd, "generator")
+	if len(generatorName) == 0 && opts != nil {
```
I would be careful with this carry, with the changes I've introduced. Because current code allows nice fallback to old behavior if no new resources are available. I'd rather see this being hooked into the discover logic below, imho.

liggitt replied:
Discovery doesn't find origin resources today. Our defaults exist on every supported version (dc and pod).

soltysh commented:
> @soltysh - extended test namespace creation

That part LGTM.

liggitt commented:
@deads2k use of discovery for core controllers (like the namespace controller) requires ensuring a role and rolebinding that allows that is present at server start. added the ability to scope a reconcile-cluster-role-bindings invocation to bindings related to named roles, and added auto-reconcile for a discovery role and rolebinding at start

deads2k commented on pkg/cmd/admin/policy/reconcile_clusterrolebindings.go:
```diff
@@ -193,6 +216,11 @@ func (o *ReconcileClusterRoleBindingsOptions) ChangedClusterRoleBindings() ([]*a
 
+	if len(rolesNotFound) != 0 {
+		// return the known changes and the error so that a caller can decide if he wants a partial update
+		return changedRoleBindings, fmt.Errorf("did not find requested cluster role %s", rolesNotFound.List())
```
You need to update callers (`RunReconcileClusterRoleBindings` comes to mind), to handle partial success.

deads2k commented on pkg/cmd/server/bootstrappolicy/constants.go:
```diff
@@ -107,6 +108,7 @@ const (
 	SDNReaderRoleBindingName         = SDNReaderRoleName + "s"
 	SDNManagerRoleBindingName        = SDNManagerRoleName + "s"
 	WebHooksRoleBindingName          = WebHooksRoleName + "s"
+	DiscoveryRoleBindingName         = DiscoveryRoleName + "-binding"
```
`system-discoverers`?

deads2k commented on pkg/serviceaccounts/client.go:
```diff
@@ -66,7 +65,10 @@ func Clients(config restclient.Config, tokenRetriever TokenRetriever, namespace,
+	if len(config.UserAgent) > 0 {
+		config.UserAgent += " "
+	}
+	config.UserAgent += fmt.Sprintf("system:serviceaccount:%s:%s", namespace, name)
```
This looks suspiciously like `MakeUsername`

deads2k commented on pkg/cmd/server/kubernetes/node_config.go:
```diff
@@ -187,9 +187,6 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
 		glog.Warningf("Using FakeOOMAdjuster for docker-in-docker compatibility")
 		cfg.OOMAdjuster = oom.NewFakeOOMAdjuster()
-		glog.Warningf("Disabling cgroup manipulation of nested docker daemon for docker-in-docker compatibility")
-		cfg.RuntimeCgroups = ""
```
This has been both added and removed in your rebase.  Intentional?

liggitt replied:
Yes, field name changed, but default also changed to empty, so it wasn't
needed.



On Wed, Mar 16, 2016 at 8:03 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/kubernetes/node_config.go
> <https://github.com/openshift/origin/pull/7975#discussion_r56321794>:
>
> > @@ -187,9 +187,6 @@ func BuildKubernetesNodeConfig(options configapi.NodeConfig) (*NodeConfig, error
> >  	if value := cmdutil.Env("OPENSHIFT_DIND", ""); value == "true" {
> >  		glog.Warningf("Using FakeOOMAdjuster for docker-in-docker compatibility")
> >  		cfg.OOMAdjuster = oom.NewFakeOOMAdjuster()
> > -		glog.Warningf("Disabling cgroup manipulation of nested docker daemon for docker-in-docker compatibility")
> > -		cfg.RuntimeCgroups = ""
>
> This has been both added and removed in your rebase. Intentional?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7975/files/6401e0e510ff9c48b250a7d23444f768dc7be254..a3ba63351a644d8d81c32a6f624cc37bd82e05ea#r56321794>
>

deads2k commented on pkg/cmd/server/origin/master.go:
```diff
@@ -558,7 +559,13 @@ func (c *MasterConfig) InstallUnprotectedAPI(container *restful.Container) ([]st
-	versionHandler := apiserver.APIVersionHandler(kapi.Codecs, versions...)
+	versionHandler := apiserver.APIVersionHandler(kapi.Codecs, func(req *restful.Request) *unversioned.APIVersions {
+		apiVersionsForDiscovery := unversioned.APIVersions{
+			// TODO: ServerAddressByClientCIDRs: s.getServerAddressByClientCIDRs(req.Request),
```
Seems like something we'll want.  Followup issue?

liggitt replied:
Already on the follow up list

On Wed, Mar 16, 2016 at 8:04 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/server/origin/master.go
> <https://github.com/openshift/origin/pull/7975#discussion_r56321878>:
>
> > @@ -558,7 +559,13 @@ func (c *MasterConfig) InstallUnprotectedAPI(container *restful.Container) ([]st
> >
> >  // initAPIVersionRoute initializes the osapi endpoint to behave similar to the upstream api endpoint
> >  func initAPIVersionRoute(root *restful.WebService, prefix string, versions ...string) {
> > -	versionHandler := apiserver.APIVersionHandler(kapi.Codecs, versions...)
> > +	versionHandler := apiserver.APIVersionHandler(kapi.Codecs, func(req *restful.Request) *unversioned.APIVersions {
> > +		apiVersionsForDiscovery := unversioned.APIVersions{
> > +			// TODO: ServerAddressByClientCIDRs: s.getServerAddressByClientCIDRs(req.Request),
>
> Seems like something we'll want. Followup issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7975/files/9ad61a43602f99e9be4d2504b48a13352980ea49..6401e0e510ff9c48b250a7d23444f768dc7be254#r56321878>
>

deads2k commented:
Hrm, yeah, it's added some logic to hide comments on diffs that it thinks are out date, but that's not what I want it to do.

deads2k commented:
Minor comments, I'm willing to take them all as close followups.

I'd like a squash if you get to it before the [merge] job starts.

liggitt commented:
Hold merging, the job validation upstream changed last night. I'll update and squash

liggitt commented:
[testonlyextended][extended:all]

liggitt commented:
fixed up upstream and squashed.
[merge]

liggitt commented:
[vipmerge]

soltysh commented:
Yay :tada:

liggitt commented:
Probably the TLS etcd unit tests. Will update with the same workaround we did for the f5 tests to remove ecdhe ciphers

liggitt commented:
Re[vipmerge] in the meantime

liggitt commented:
Updated to run the master pkg unit tests (which spawn lots of storage connections) using the old http etcd test instance. [vipmerge]

fabianofranz commented:
![10nhmcq0yooi3c](https://cloud.githubusercontent.com/assets/158611/13824193/be003b70-eb8b-11e5-8319-c2b75b37abb3.gif)

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/run.go:
```diff
@@ -69,7 +69,15 @@ kubectl run nginx --image=nginx --command -- <cmd> <arg1> ... <argN>
 kubectl run pi --image=perl --restart=OnFailure -- perl -Mbignum=bpi -wle 'print bpi(2000)'`
 )
 
+type RunOptions struct {
```
Did this ever make it upstream?

soltysh replied:
@liggitt might know?
